consensus/p2p/validator/transition.go

```diff
@@ -3,159 +3,296 @@ package validator
-		ValidRound: state.ValidRound,
-		Value:      state.Value,
-	}, nil
+	mempoolTransactions := make([]mempool.BroadcastedTransaction, len(transactions))
+	for i := range mempoolTransactions {
+		mempoolTransactions[i] = mempool.BroadcastedTransaction{
+			Transaction:   transactions[i].Transaction,
+			DeclaredClass: transactions[i].Class,
+			PaidFeeOnL1:   transactions[i].PaidFeeOnL1,
+		}
+	}
+
+	if err := t.builder.RunTxns(state.BuildState, mempoolTransactions); err != nil {
+		return nil, err
+	}
+
+	return state, nil
 }
 
 // TODO: Implement this function properly
 func (t *transition) OnProposalCommitment(
 	ctx context.Context,
 	state *ReceivingTransactionsState,
-	commitment *consensus.ProposalCommitment,
+	commitment *types.ProposalCommitment,
 ) (*AwaitingProposalFinState, error) {
+	buildResult, err := t.finishAndValidateBlock(state.BuildState, commitment)
+	if err != nil {
+		return nil, err
+	}
+
 	return &AwaitingProposalFinState{
-		Header:     state.Header,
-		ValidRound: state.ValidRound,
-		Value:      state.Value,
+		Proposal: &starknet.Proposal{
+			MessageHeader: *state.Header,
+			Value:         (*starknet.Value)(buildResult.Pending.Block.Hash),
+			ValidRound:    state.ValidRound,
+		},
+		BuildResult: &buildResult,
 	}, nil
 }
 
 // TODO: Implement this function properly
 func (t *transition) OnProposalFin(
 	ctx context.Context,
 	state *AwaitingProposalFinState,
-	fin *consensus.ProposalFin,
+	fin *types.ProposalFin,
 ) (*FinState, error) {
 	// Check if expected and actual are both nil or both non-nil
-	if (state.Value == nil) != (fin.ProposalCommitment == nil) {
+	if (state.Proposal.Value == nil) != (fin == nil) {
 		return nil, errHashMismatch
 	}
 
 	// If both are non-nil, check if they are equal
-	if state.Value != nil {
-		expected := state.Value.Hash()
-		actual := starknet.Hash(felt.FromBytes(fin.ProposalCommitment.Elements))
+	if state.Proposal.Value != nil {
+		expected := state.Proposal.Value.Hash()
+		actual := starknet.Hash(*fin)
 		if expected != actual {
 			return nil, errHashMismatch
 		}
 	}
 
-	return &FinState{
-		MessageHeader: *state.Header,
-		Value:         state.Value,
-		ValidRound:    state.ValidRound,
-	}, nil
+	return (*FinState)(state), nil
+}
+
+func (t *transition) finishAndValidateBlock(
+	buildState *builder.BuildState,
+	receivedCommitment *types.ProposalCommitment,
+) (builder.BuildResult, error) {
+	buildResult, err := t.builder.Finish(buildState)
+	if err != nil {
+		return builder.BuildResult{}, err
+	}
+
+	computed, err := buildResult.ProposalCommitment()
+	if err != nil {
+		return builder.BuildResult{}, err
+	}
+
+	if err := compareProposalCommitment(&computed, receivedCommitment); err != nil {
+		return builder.BuildResult{}, err
+	}
+
+	return buildResult, nil
+}
+
+// Todo: the validator interface assumes that the msgs are prevalidated before it is called.
+// This is an issue because proto3 may leave messages empty, whereas starknet requires them
+// to be present
```

Comment:
Is this Todo still relevant and should it be above `compareFeltField`?

Reply:
Yes, the TODO was because this function is expecting that `received` is non-nil`, and the caller needs to handle nil (possible, because of protobuf) before calling this function.